Add Singapore Public Holiday

captainjaypee01 opened the pull request:
### Singapore Holidays

- New Year's Day
- Chinese New Year
- Good Friday
- Hari Raya Puasa
- Labour Day
- Vesak Day
- Hari Raya Haji
- National Day
- Deepavali
- Christmas Day

## Changes
- Replacement for `Holidays`. If falls on `Sunday`, `Monday `will be the replacement. For `Chinese New Year`, if the Day is Sunday and Monday then the `Tuesday` will be the replacement for `Sunday`.
- Used `IntlDateFormatter` for accurate calendar calculations
- Added `Chinese Calendar` to get the Chinese New Year and Vesak Day dates
- Added `Islamic Calendar` to get the Hari Raya Puasa and Hari Raya Haji dates
- Added `Indian Calendar` to get the Kartik month which is 8th month of the Indian Calendar and 10th month of the Gregorian Calendar.
- Used the `solaris/php-moon-phase` for accurate computation of moon phases. Especially for `Deepavali` holiday, as this holiday is based on moon appearance

## Additional Notes

1. The `solaris/php-moon-phase` extension is required for this addition. It is referenced in the `composer.json` file.

## References:
Singapore's Ministry of Manpower to reference the Holidays in Singapore
[MOM Gov Public Holiday](https://www.mom.gov.sg/employment-practices/public-holidays)

Deepavali date is based on the appearance of the moon.
[Deepavali is subject to change](https://www.mom.gov.sg/faq/public-holidays/will-deepavali-date-be-subject-to-change)

[Chinese Calendar](https://en.wikipedia.org/wiki/Chinese_calendar)
[Islamic Calendar](https://www.muis.gov.sg/Media/Islamic-Calendar)
[Indian Calendar](https://en.wikipedia.org/wiki/Hindu_calendar)

captainjaypee01 commented:
I added REPLACEMENT feature when holiday is in Sunday

captainjaypee01 commented:
Refactor replacement for Sunday holiday. Trying to think if it can implemented in the Holiday class but it seems that not all countries are doing the same process

captainjaypee01 commented:
I added the Deepavali Holiday using the moon phases. I used the "solaris/php-moon-phase" package to determine the phases of the moon. I based the date using the Indian Calendar in Kartik month

captainjaypee01 commented:
Rebased to the latest main branch

Nielsvanpach commented on src/Countries/Singapore.php:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    /** @return array<string, string|CarbonImmutable> */
+    protected function replacementForSunday(string $name, int $year, string $input, int $addDays = 0): array
```
Could you make this name & code more readable, so it's immediately obvious what it does?

captainjaypee01 replied:
I will review on this.

Nielsvanpach commented on src/Countries/Singapore.php:
```diff
@@ -0,0 +1,171 @@
+        return (int) $hijriYear;
+    }
+
+    protected function chineseCalendar(string $input, int $year): string
```
I think you could use the ChineseCalendar trait

captainjaypee01 replied:
I will review on this.

Nielsvanpach commented on src/Countries/Singapore.php:
```diff
@@ -0,0 +1,171 @@
+    }
+
+
+    protected function getIslamicFormatter(): IntlDateFormatter
```
@thecaliskan could you review the Islamic part?

thecaliskan replied:
@Nielsvanpach I will prepare a script and check it when I am available.

captainjaypee01 replied:
Hi @thecaliskan and @Nielsvanpach I have updated my changes in this branch. You may want to check 

Thanks

Nielsvanpach reviewed (commented):
We have a ChineseCalendar trait you can use: https://github.com/spatie/holidays/blob/main/src/Calendars/ChineseCalendar.php

For the Islamic & Indian calendars, could you use the same reusable approach as the ChineseCalendar trait? Thanks!

captainjaypee01 commented:
Hi @Nielsvanpach, I have added the Islamic Calendar and Indian Calendar and I used the Chinese Calendar Trait.

Regarding Deepavali Date, I didn't include this holiday in the Indian Calendar. I was thinking of adding it but I'm not sure if all countries celebrating the Deepavali Date is the same with Singapore.

For the replacement function I renamed it to "`replaceHolidayFallsOnSunday`". This function is more on replacing the holidays in Singapore if it falls on Sunday. But if there's a consecutive holidays like in Chinese New Year which falls on Sunday and Monday , the Sunday will move to Tuesday as the replacement. I think not all countries doing this scenario.

captainjaypee01 commented:
Hi @Nielsvanpach , I'm planning to add a function to check if the holiday is a regular or a replacement holiday.

captainjaypee01 commented:
Hi @Nielsvanpach , is there anything that I need to add on this PR?

Nielsvanpach commented on composer.json:
```diff
@@ -22,7 +22,9 @@
-        "nesbot/carbon": "^2.72.1|^3.0"
+        "nesbot/carbon": "^2.72.1|^3.0",
+        "ext-calendar": "*",
+        "solaris/php-moon-phase": "^2.1"
```
solaris/php-moon-phase can be removed, I think?

captainjaypee01 replied:
I think I used the moon phase for deepavali event as this holiday is based on the moon phases

Nielsvanpach commented on src/Countries/Singapore.php:
```diff
@@ -0,0 +1,112 @@
+     * 
+     * @return array<string, string|CarbonImmutable> 
+     */
+    protected function replaceHolidayFallsOnSunday(string $holidayName, int $year, string $input, int $moveDays = 1): array
```
I think you could use the Observable trait for this

captainjaypee01 replied:
I will convert this to Observable trait

Nielsvanpach reviewed (changes requested):
Looks good!
I'm not sure if the IslamicCalendar gives proper results, as it was previously stated this is different from country to country.
